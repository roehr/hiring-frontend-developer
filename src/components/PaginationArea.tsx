import * as React from "react";
import {Pagination} from "../testimonial.model";
import {computePaginationList} from "../utils";

interface PaginationProps {
    pagination: Pagination
    onPageClicked: (page: number) => void
}
const PaginationArea = (props: PaginationProps) => {

    const onPreviousClicked = () => {
        if(props.pagination.current_page !== 1) {
            props.onPageClicked(props.pagination.current_page-1)
        }
    }

    const onNextClicked = () => {
        if(props.pagination.current_page < props.pagination.total_pages) {
            props.onPageClicked(props.pagination.current_page+1)
        }
    }

    const onPageNumberClicked = (candidate:string) => {
        if(candidate!=="...") {
            props.onPageClicked(Number(candidate))
        }
    }


    const renderPages = () =>
       computePaginationList(props.pagination.current_page, props.pagination.total_pages).map(
           (value, index)=>
               <div key={index} onClick={()=>{onPageNumberClicked(value)}}  data-testid={`pagination-button${Number(value) === props.pagination.current_page ? '-active': index}`}>{value}</div>

       )


    return <div className="pagination">
        <div className={`previous ${props.pagination.current_page !== 1 ? '' : 'disabled'}`} onClick={onPreviousClicked}  data-testid="prev-button">Previous</div>
        <div className="pages">
            {renderPages()}
        </div>
        <div className={`next ${props.pagination.current_page !== 1 ? '' : 'disabled'}`} onClick={onNextClicked}  data-testid="next-button">Next</div>
    </div>
}

export {PaginationArea}