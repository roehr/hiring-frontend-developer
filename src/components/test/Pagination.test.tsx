import {PaginationArea} from "../PaginationArea";
import {render} from "@testing-library/react";
import React from "react";
import userEvent from "@testing-library/user-event";

describe('Pagination test', () => {
    test('renders Pagination', () => {
        const underTest = render(<PaginationArea pagination={{current_page:2, total_pages:10, total_count:200}} onPageClicked={jest.fn}/>)
        expect(underTest).toMatchSnapshot();
    });
    test('Only one Page', async () => {
        const pagination = {current_page:1, total_pages:1, total_count:20}
        const underTest = render(<PaginationArea pagination={pagination} onPageClicked={jest.fn()}/>)
        expect(underTest.getByTestId("prev-button").className).toContain("disabled")
        expect(underTest.getByTestId("next-button").className).toContain("disabled")
    })
    describe('Clicks', () => {
        const onClick = jest.fn();
        const pagination = {current_page:4, total_pages:10, total_count:200};
        test('Prev Button', async () => {
            const underTest = render(<PaginationArea pagination={pagination} onPageClicked={onClick}/>)
            await userEvent.click(underTest.getByTestId("prev-button"))
            expect(onClick).toHaveBeenCalled()
        });
        test('Next Button', async () => {
            const underTest = render(<PaginationArea pagination={pagination} onPageClicked={onClick}/>)
            await userEvent.click(underTest.getByTestId("next-button"))
            expect(onClick).toHaveBeenCalled()
        });
        test('Click on a Page Button', async () => {
            const underTest = render(<PaginationArea pagination={pagination} onPageClicked={onClick}/>)
            await userEvent.click(underTest.getByTestId("pagination-button1"))
            expect(onClick).toHaveBeenCalled()
        })
    });
})

