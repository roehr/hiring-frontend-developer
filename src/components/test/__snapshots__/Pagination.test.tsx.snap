// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination test renders Pagination 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pagination"
      >
        <div
          class="previous "
          data-testid="prev-button"
        >
          Previous
        </div>
        <div
          class="pages"
        >
          <div
            class="pagination-button"
            data-testid="pagination-button0"
          >
            1
          </div>
          <div
            class="pagination-button"
            data-testid="pagination-button-active"
          >
            2
          </div>
          <div
            class="pagination-button"
            data-testid="pagination-button2"
          >
            3
          </div>
          <div
            class="pagination-button"
            data-testid="pagination-button3"
          >
            ...
          </div>
          <div
            class="pagination-button"
            data-testid="pagination-button4"
          >
            8
          </div>
          <div
            class="pagination-button"
            data-testid="pagination-button5"
          >
            9
          </div>
          <div
            class="pagination-button"
            data-testid="pagination-button6"
          >
            10
          </div>
        </div>
        <div
          class="next "
          data-testid="next-button"
        >
          Next
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pagination"
    >
      <div
        class="previous "
        data-testid="prev-button"
      >
        Previous
      </div>
      <div
        class="pages"
      >
        <div
          class="pagination-button"
          data-testid="pagination-button0"
        >
          1
        </div>
        <div
          class="pagination-button"
          data-testid="pagination-button-active"
        >
          2
        </div>
        <div
          class="pagination-button"
          data-testid="pagination-button2"
        >
          3
        </div>
        <div
          class="pagination-button"
          data-testid="pagination-button3"
        >
          ...
        </div>
        <div
          class="pagination-button"
          data-testid="pagination-button4"
        >
          8
        </div>
        <div
          class="pagination-button"
          data-testid="pagination-button5"
        >
          9
        </div>
        <div
          class="pagination-button"
          data-testid="pagination-button6"
        >
          10
        </div>
      </div>
      <div
        class="next "
        data-testid="next-button"
      >
        Next
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
